tensorboard/data/server/reservoir.rs

```diff
@@ -0,0 +1,329 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+//! Reservoir sampling with preemption and deferred commits.
+
+use rand::{
+    distributions::{Distribution, Uniform},
+    Rng, SeedableRng,
+};
+use rand_chacha::ChaCha20Rng;
+
+type Step = i64;
+
+/// A [reservoir sampling] data structure, with support for preemption and deferred "commits" of
+/// records to a separate destination for better concurrency. This structure always keeps the
+/// latest record in the reservoir, and therefore must inspect every record in the stream.
+///
+/// **Note:** Preemption support is not yet implemented.
+///
+/// # Preemption
+///
+/// All records stored in this reservoir have a *step* and a *payload*. The step, a non-negative
+/// integer, is expected to be monotonically (strictly) increasing over time. Whenever a new record
+/// arrives with step `s`, it preempts any records with steps not smaller than `s`.
```

Comment:
Updated docs.

Reply:
Got it.  Can we add the "non-preempted" qualifier to the `self.seen` docs to match what you wrote here?   As in "total number of _non-preempted_ records passed in the stream" - having that there is helpful to make it clear up-front that we're attempting to model the record steam with preemptions removed.

As discussed, SGTM to implement the behavior as you describe that's closer to a uniform sample, modulo checking that the logic introduced in the next PR resolves the "preempt _all_ the records" case by setting `self.seen` back to 0 and effectively returning to square one, since that means for that particular special case we can actually start eagerly filling the reservoir again without distortion.

Reply:
Thanks; yes. Done.